src/CFamily/Rules/RulesConfigFixup.cs

```diff
@@ -0,0 +1,198 @@
+            { "PPIncludeNotAtTop","S954" },
+            { "PPIncludeTime", "S991" },
+            { "TrigraphUsage", "S797" },
+            { "ContinueUsage", "S909" },
+            { "LineLength", "S103" },
+            { "FileLoc", "S104" },
+            { "GotoUsage", "S907" },
+            { "IdentifierLongerThan31", "S799" },
+            { "GlobalNamespaceMembers", "S997" },
+            { "PPIncludeNonStandardCharacters","S955" },
+            { "BackJumpWithGoto", "S999" },
+            { "FileComplexity", "S1908" },
+            { "TabCharacter", "S105" },
+            { "DigraphUsage", "S798" },
+            { "InvalidEscapeSequence", "S796" },
+            { "ObsoletePosixFunction", "S1911" },
+            { "PPIncludeSignal", "S987" },
+            { "PPBackslashNotLastCharacter", "S1916" },
+            { "ClassComplexity", "S1311" },
+            { "SwitchLabelPlacement", "S916" },
+            { "PPIncludeStdio", "S988" },
+            { "FunctionComplexity", "S1541" },
+            { "CommentMixedStyles", "S1917" },
+            { "OneStatementPerLine", "S122" },
+            { "CommaAndOrOverloaded", "S919" },
+            { "CommentedCode", "S125" },
+            { "FunctionSinglePointOfExit", "S1005" },
+            { "PPIncludeCHeader","S1051" },
+            { "EnumPartialInitialization", "S841" },
+            { "UnaryAndOverloaded", "S877" },
+            { "ParsingError", "S2260" },
+            { "SwitchWithoutDefault", "S131" },
+            { "PPStringifyAndPastingUsage", "S968" },
+            { "PPUndefUsage", "S959" },
+            { "ClassName", "S101" },
+            { "EmptyCompoundStatement", "S108" },
+            { "PPDefineOrUndefFromBlock","S958" },
+            { "PPBadIncludeForm", "S956" }
+        };
+
+        private static readonly IReadOnlyDictionary<string, string> fullLegacyToNewKeyMap = CalculateFullKeyMap();
+
+        private static IReadOnlyDictionary<string, string> CalculateFullKeyMap()
+        {
+            var mapWithLanguagePrefixes = new Dictionary<string, string>();
+            foreach (var partial in partialLegacyToNewKeyMap)
+            {
+                mapWithLanguagePrefixes[$"{SonarLanguageKeys.C}:{partial.Key}"] = $"{SonarLanguageKeys.C}:{partial.Value}";
+                mapWithLanguagePrefixes[$"{SonarLanguageKeys.CPlusPlus}:{partial.Key}"] = $"{SonarLanguageKeys.CPlusPlus}:{partial.Value}";
+            }
+            return mapWithLanguagePrefixes;
+        }
+
+        private readonly ILogger logger;
+
+        public RulesConfigFixup(ILogger logger) => this.logger = logger ?? throw new System.ArgumentNullException(nameof(logger));
+
+        /// <summary>
+        /// Translates any legacy rule keys in the input to new Sxxx rule keys
+        /// </summary>
+        public RulesSettings Apply(RulesSettings input)
+        {
+            /*
+
+            We're making a shallow copy of the list of rules. If we modify the original list, any exclusions we 
```

Comment:
The first part of this comment giving the reasoning is taken from here: https://github.com/SonarSource/sonarlint-visualstudio/blob/63a9400ef82f8e62298b8ab2eb433723dc17a8f8/src/Core/CFamily/DynamicCFamilyRulesConfig.cs#L201

I've added more details about the specific legacy key scenarios below.